### PR TITLE
New package: Canon.i-SENSYS.MF4010-MF4018 version 2.10

### PR DESCRIPTION
**Pull request by rakleed**
- [ ] Have you signed the [Contributor License Agreement](https://cla.opensource.microsoft.com/microsoft/winget-pkgs)?
- [ ] Have you checked that there aren't other open [pull requests](https://github.com/microsoft/winget-pkgs/pulls) for the same manifest update/change?
- [ ] This PR only modifies one (1) manifest
- [ ] Have you [validated](https://github.com/microsoft/winget-pkgs/blob/master/AUTHORING_MANIFESTS.md#validation) your manifest locally with `winget validate --manifest <path>`?
- [ ] Have you tested your manifest locally with `winget install --manifest <path>`?
- [ ] Does your manifest conform to the [1.5 schema](https://github.com/microsoft/winget-pkgs/tree/master/doc/manifest/schema/1.5.0)?

Note: `<path>` is the name of the directory containing the manifest you're submitting.

---

 ###### Microsoft Reviewers: [Open in CodeFlow](https://microsoft.github.io/open-pr/?codeflow=https://github.com/microsoft/winget-pkgs/pull/132956)

### CODE REVIEW

**Comment by rakleed** [comment]
https://www.canon-europe.com/support/consumer/products/printers/i-sensys/mf-series/i-sensys-mf4018.html?type=drivers&detailId=tcm:13-997507&os=Windows%2010%20(64-bit)&language=en&productTcmUri=tcm:13-728885

**Comment by stephengillie** [comment]
Hi @rakleed,

Could you add an InstallerUrl?

(Automated response - build 528.)

**Comment by rakleed** [comment]
@stephengillie

**Comment by stephengillie** [comment]
Hi @rakleed,

I can't determine that these drivers are installed on the system - they don't show up in Apps Settings nor could I find any files named after the package. How should I be interacting with or verifying this package?

**Comment by rakleed** [comment]
The downloaded exe file is a self-extracting archive, inside of which there is a folder MF4010_MFDrivers_W64_uk_EN\x64, where the Setup.exe file is located, which actually installs the drivers. I'll try to ask Canon support if it is possible to run not only unpacking, but also driver installation with one command.

**Comment by stephengillie** [comment]
Ah, you're describing a Nested Installer. Here is the [schema](https://github.com/microsoft/winget-pkgs/blob/master/doc/manifest/schema/1.5.0/installer.md), which has a little more info. Basically we'd add something like: 

```
    NestedInstallerType: exe (or update if different)
    NestedInstallerFiles: Setup.exe
      - RelativeFilePath: /
```

File: manifests/c/Canon/i-SENSYS/MF4010-MF4018/2.10/Canon.i-SENSYS.MF4010-MF4018.installer.yaml [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+PackageIdentifier: Canon.i-SENSYS.MF4010-MF4018
+PackageVersion: "2.10"
+InstallerType: exe
+InstallerSwitches:
+  Silent: /qn
+  SilentWithProgress: /qn
+Installers:
+- Architecture: x64
+  InstallerUrl: https://gdlp01.c-wss.com/gds/5/0100004725/03/MF4010_MFDrivers_W64_uk_EN.exe
+  InstallerSha256: C5E0C5800DE336A886D3A00666C6FE63349F806C56D4C5760C7CCFC2914A89AA
```
Comment (stephengillie) [inline manifests/c/Canon/i-SENSYS/MF4010-MF4018/2.10/Canon.i-SENSYS.MF4010-MF4018.installer.yaml:13]:
```suggestion
  InstallerSha256: C5E0C5800DE336A886D3A00666C6FE63349F806C56D4C5760C7CCFC2914A89AA
    NestedInstallerType: exe
    NestedInstallerFiles: Setup.exe
      - RelativeFilePath: /
```
These are example values - please verify before confirming.

**Comment by rakleed** [comment]
Maybe the problem with `RelativeFilePath`?

File: manifests/c/Canon/i-SENSYS/MF4010-MF4018/2.10/Canon.i-SENSYS.MF4010-MF4018.installer.yaml [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+  NestedInstallerType: exe
+  NestedInstallerFiles: Setup.exe
+    - RelativeFilePath: x64
```
Comment (stephengillie) [inline manifests/c/Canon/i-SENSYS/MF4010-MF4018/2.10/Canon.i-SENSYS.MF4010-MF4018.installer.yaml:16]:
```suggestion
InstallerType: zip
InstallerSwitches:
  Silent: /qn
  SilentWithProgress: /qn
Installers:
- Architecture: x64
  InstallerUrl: https://gdlp01.c-wss.com/gds/5/0100004725/03/MF4010_MFDrivers_W64_uk_EN.exe
  InstallerSha256: C5E0C5800DE336A886D3A00666C6FE63349F806C56D4C5760C7CCFC2914A89AA
  NestedInstallerType: exe
  NestedInstallerFiles: 
    - RelativeFilePath: x64\Setup.exe
```

Comment (stephengillie) [reply manifests/c/Canon/i-SENSYS/MF4010-MF4018/2.10/Canon.i-SENSYS.MF4010-MF4018.installer.yaml:16]: I tried this a few times to figure out the right pattern. First, my earlier suggestion wasn't actually valid YAML. I believe `NestedInstallerFiles` shouldn't have any text following it, to represent an array field. So moving the installer name into the `RelativeFilePath` parameter allows the package manager to parse this section. 

On most runs, the installer switches were applied to the outer EXE. This caused only empty directories to become extracted, intsead of the driver files. A feature here is that installer switches are supposed to run against the nested installer, not the outer installer. Based on this, I'm not sure if we support EXE within EXE. 

The installer is a self-extracting ZIP, so I guessed that treating it as one might allow the inner installer to be handled correctly. And this seems to be the case. Manual Validation is still underway, but the driver files appear to have extracted correctly, and shows up in Apps Settings: 
image

**Comment by rakleed** [comment]
@stephengillie

**Comment by stephengillie** [comment]
Automatic Validation ended with:
>  Installation failed with exit code -1978335226 2024-01-10 16:22:13.427 [CLI ] ShellExecute installer failed: 4294967295

(Automated response - build 549.)

**Review by stephengillie — APPROVED** [review]

**Pull request merged**